File: packages/cli/src/templates/cc-native/_cc-native/lib/reviewers/codex.py
"""
CC-Native Codex Reviewer Module.

Runs Codex CLI to review plans.
"""

import json
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import Any, Dict

# Import from parent lib
_lib_dir = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(_lib_dir))

from utils import ReviewerResult, eprint, parse_json_maybe, coerce_to_review
from .base import REVIEW_PROMPT_PREFIX


def run_codex_review(
    plan: str,
    schema: Dict[str, Any],
    settings: Dict[str, Any],
) -> ReviewerResult:
    """Run Codex CLI to review the plan.

    Args:
        plan: The plan content to review
        schema: JSON schema for the review output
        settings: Codex reviewer settings (timeout, model)

    Returns:
        ReviewerResult with the review output
    """
    codex_settings = settings.get("reviewers", {}).get("codex", {})
    timeout = codex_settings.get("timeout", 120)
    model = codex_settings.get("model", "")

    codex_path = shutil.which("codex")
    if codex_path is None:
        eprint("[codex] CLI not found on PATH")
        return ReviewerResult(
            name="codex",
            ok=False,
            verdict="skip",
            data={},
            raw="",
            err="codex CLI not found on PATH",
        )

    eprint(f"[codex] Found CLI at: {codex_path}")

    prompt = f"""{REVIEW_PROMPT_PREFIX}
Return ONLY a JSON object that matches this JSON Schema:
{json.dumps(schema, ensure_ascii=False)}

PLAN:
<<<
{plan}
>>>
"""

    with tempfile.TemporaryDirectory() as td:
        td_path = Path(td)
        schema_path = td_path / "schema.json"
        out_path = td_path / "codex_review.json"

        schema_path.write_text(json.dumps(schema, indent=2), encoding="utf-8")

        cmd = [
            codex_path,
            "exec",
            "--full-auto",
            "--sandbox",
            "read-only",
            "--output-schema",
            str(schema_path),
            "-o",
            str(out_path),
            "-",
        ]

        if model:
            cmd.insert(2, "--model")
            cmd.insert(3, model)

        eprint(f"[codex] Running command: {' '.join(cmd)}")

        try:
            p = subprocess.run(
                cmd,
                input=prompt,
                text=True,
                capture_output=True,
                timeout=timeout,
                encoding='utf-8',
                errors='replace',
            )
        except subprocess.TimeoutExpired:
            eprint(f"[codex] TIMEOUT after {timeout}s")
            return ReviewerResult("codex", False, "error", {}, "", f"codex timed out after {timeout}s")
        except Exception as ex:
            eprint(f"[codex] EXCEPTION: {ex}")
            return ReviewerResult("codex", False, "error", {}, "", f"codex failed to run: {ex}")

        eprint(f"[codex] Exit code: {p.returncode}")

        raw = ""
        if out_path.exists():
            raw = out_path.read_text(encoding="utf-8", errors="replace")

        obj = parse_json_maybe(raw) or parse_json_maybe(p.stdout)
        ok, verdict, norm = coerce_to_review(obj, "Retry or check CLI auth/config.")

        err = (p.stderr or "").strip()
        return ReviewerResult("codex", ok, verdict, norm, raw or p.stdout, err)

File: packages/cli/src/templates/cc-native/_cc-native/lib-ts/reviewers/codex.ts
/**
 * Codex CLI plan reviewer.
 * Runs Codex in non-interactive mode with read-only sandbox.
 * See cc-native-plan-review-spec.md §4.11
 */

import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import { logDebug, logInfo, logWarn, logError } from "../../../_shared/lib-ts/base/logger.js";
import { findExecutable, execFileAsync, getInternalSubprocessEnv } from "../../../_shared/lib-ts/base/subprocess-utils.js";
import { parseJsonMaybe, coerceToReview } from "../json-parser.js";
import { REVIEW_PROMPT_PREFIX } from "../types.js";
import type { ReviewerResult, ReviewOptions } from "../types.js";
import { makeResult } from "./types.js";
import type { Reviewer } from "./types.js";

/**
 * Codex reviewer — runs codex exec --sandbox read-only.
 */
export class CodexReviewer implements Reviewer {
  private settings: Record<string, unknown>;

  constructor(settings: Record<string, unknown>) {
    this.settings = settings;
  }

  async review(
    plan: string,
    schema: Record<string, unknown>,
    options: ReviewOptions,
  ): Promise<ReviewerResult> {
    return runCodexReview(plan, schema, this.settings);
  }
}

/**
 * Run Codex CLI to review the plan.
 * Never throws — returns error ReviewerResult on failure.
 */
export async function runCodexReview(
  plan: string,
  schema: Record<string, unknown>,
  settings: Record<string, unknown>,
): Promise<ReviewerResult> {
  const codexSettings =
    ((settings.reviewers as Record<string, unknown> | undefined)?.codex as
      | Record<string, unknown>
      | undefined) ?? {};
  const timeout = (codexSettings.timeout as number) ?? 120;
  const model = (codexSettings.model as string) ?? "";

  const codexPath = findExecutable("codex");
  if (!codexPath) {
    logWarn("codex", "CLI not found on PATH");
    return makeResult("codex", false, "skip", {}, "", "codex CLI not found on PATH");
  }

  logDebug("codex", `Found CLI at: ${codexPath}`);

  const prompt = `${REVIEW_PROMPT_PREFIX}
Return ONLY a JSON object that matches this JSON Schema:
${JSON.stringify(schema)}

PLAN:
<<<
${plan}
>>>
`;

  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "codex-review-"));

  try {
    const schemaPath = path.join(tmpDir, "schema.json");
    const outPath = path.join(tmpDir, "codex_review.json");

    fs.writeFileSync(schemaPath, JSON.stringify(schema, null, 2), "utf-8");

    const cmdArgs = ["exec", "--sandbox", "read-only"];

    if (model) {
      cmdArgs.push("--model", model);
    }

    cmdArgs.push("--output-schema", schemaPath, "-o", outPath, "-");

    logDebug("codex", `Running command: codex ${cmdArgs.join(" ")}`);

    const env = getInternalSubprocessEnv();

    const result = await execFileAsync(codexPath, cmdArgs, {
      input: prompt,
      timeout: timeout * 1000,
      env: env as Record<string, string>,
      maxBuffer: 10 * 1024 * 1024,
      shell: process.platform === "win32",
    });

    if (result.killed || result.signal === "SIGTERM") {
      logWarn("codex", `TIMEOUT after ${timeout}s`);
      return makeResult("codex", false, "error", {}, "", `codex timed out after ${timeout}s`);
    }

    const stderrText = result.stderr.trim();

    // Log exit code and stderr tail for ALL non-zero exits (aids diagnosis of intermittent failures)
    if (result.exitCode !== 0) {
      const stderrTail = stderrText.slice(-500);
      logWarn("codex", `Exited with code ${result.exitCode}, stderr_len=${stderrText.length}, stderr_tail: ${stderrTail}`);
    }

    if (!result.stdout && !stderrText && !fs.existsSync(outPath) && result.exitCode !== 0) {
      logError("codex", `Process exited with code ${result.exitCode} and no output`);
      return makeResult("codex", false, "error", {}, "", `codex failed to run (exit ${result.exitCode})`);
    }

    let raw = "";
    const outExists = fs.existsSync(outPath);
    if (outExists) {
      raw = fs.readFileSync(outPath, "utf-8");
    }

    logDebug("codex", `Exit code: ${result.exitCode}, outFile=${outExists} (${raw.length} chars), stdout=${result.stdout.length} chars`);

    const obj = parseJsonMaybe(raw) ?? parseJsonMaybe(result.stdout);
    const [ok, verdict, norm] = coerceToReview(
      obj,
      "Retry or check CLI auth/config.",
    );

    return makeResult("codex", ok, verdict, norm, raw || result.stdout, stderrText);
  } finally {
    // Clean up temp directory
    try {
      fs.rmSync(tmpDir, { recursive: true, force: true });
    } catch {
      // Best effort cleanup
    }
  }
}

